File: clp/tp/entidades/Venda.java
package clp.tp.entidades;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class Venda extends Entidade implements Totalizavel {

  private LocalDateTime dataHora;
  private List<ItemVenda> itens;

  public Venda() {
    super();
    dataHora = LocalDateTime.now();
    itens = new ArrayList<>();
  }

  public LocalDateTime getDataHora() {
    return dataHora;
  }

  public List<ItemVenda> getItens() {
    return itens;
  }

  public void adicionarItem(Produto produto, int qtd) {
    itens.add(new ItemVenda(produto, qtd));
  }

  public void removerItem(int posicao) {
    itens.remove(posicao);
  }

  public void removerItem(String nomeProduto) {
    itens.removeIf((i) -> i.getProduto().getNome().equalsIgnoreCase(nomeProduto)); // filtro com função lambda
  }

  public String toString() {

    var sb = new StringBuilder(String.format("%sData-Hora: %s\nItens:\n", super.toString(), dataHora));

    for (ItemVenda i : itens) {
      sb.append("\n  " + i.toString());
    }

    sb.append("\nTOTAL: " + total());

    return sb.toString();
  }

  @Override
  public double total() {

    var t = 0.0; // Quando se usa var, Java infere o tipo a aprtir do valor.

    for (ItemVenda i : itens) { // laço por estrutura de dados
      t += i.valor * i.qtd; // Em Java, a classe externa acessa os membros privados da classe interna. Java
                            // é fortemente tipada, mas aceita coerção por alargamento.
    }

    return t;
  }

  private class ItemVenda { // Classe interna

    private Produto produto;
    private int qtd;
    private double valor;

    public ItemVenda(Produto produto, int qtd) {
      super();
      this.produto = produto;
      this.qtd = qtd;
      this.valor = produto.getValor();
    }

    public Produto getProduto() {
      return produto;
    }

    public String toString() {
      return String.format("%15s %8.2f x %5d = %8.2f", produto.getNome(), valor, qtd, valor * qtd);
    };

  }

}

File: pmd/python/aulas/venda.py
from abc import ABC, abstractmethod

class Totalizavel(ABC):
  @abstractmethod
  def total(self): pass


class Entidade(ABC):
  def __init__(self, id):
    self.id = id

  def __str__(self):
    return f"ID: {self.id}"

class Produto(Entidade):

  def __init__(self, id, nome, valor):
    super().__init__(id)
    self.nome = nome
    self.valor = valor

  def __str__(self):
    return f"{super().__str__()}\t{self.nome}\t= {self.valor}"

class ValorInvalidoException (Exception):
  def __init__(self, valor):
    super().__init__(f"Valor inválido para produto: {valor}")


class Venda(Entidade,Totalizavel):
  def __init__(self, id):
    super().__init__(id)
    self.itens = []

  def total(self):
    s = 0.0
    for i in self.itens:
      s += i.total()
    return s 

  def __str__(self):
    s = super().__str__()
    for i in self.itens:
      s += f"\n  {i}"
    return s + f"\nTOTAL: {self.total()}"


class ItemVenda(Totalizavel):

  def __init__(self, produto, qtd=1):
    self.produto = produto
    self.qtd = qtd
    self.valor = self.produto.valor

  def total(self):
    return self.qtd * self.valor 

  def __str__(self):
    return f"{self.produto.nome}\t{self.qtd} x {self.valor} = {self.total()}"



produtos = []
while True:
  print("\n\n0 SAIR\n1 CRIAR PRODUTO\n2 IMPRIMIR")
  op = input("Digite sua opção: ")
  
  if op == "0": 
    break
  elif op == "1":
    id = int(input("ID do produto: "))
    nome = input("Nome do produto: ")
    valor = -1.0
    while valor <= 0:
      try:
        valor = float(input("Valor do produto: "))
        if valor <= 0: raise ValorInvalidoException(valor)
      except ValorInvalidoException as e:
        print(e)
    produtos.append(Produto(id, nome, valor))
  else:
    for p in produtos:
      print(p)  